tapdb/multiverse.go

```diff
@@ -283,6 +283,69 @@ func (r *rootNodeCache) wipeCache() {
 	log.Debugf("wiping universe cache")
 
 	r.allRoots.Store(&cachedRoots{})
+	r.rootIndex.Store(&rootIndex{})
+}
+
+// cachedLeafKeys is used to cache the set of leaf keys for a given universe.
+//
+// TODO(roasbeef); cacheable[T]
+type cachedLeafKeys []universe.LeafKey
+
+// Size just returns 1, as we cache based on the total number of assets, but
+// not the sum of their leaves.
+func (c *cachedLeafKeys) Size() (uint64, error) {
+	return uint64(1), nil
+}
+
+// leafKeysCache is used to cache the set of leaf keys for a given universe.
+type leafKeysCache = lru.Cache[treeID, *cachedLeafKeys]
+
+// universeLeafCaches is used to cache the set of leaf keys for a given
+// universe.
+type universeLeafCache struct {
+	leafCache *leafKeysCache
+}
+
+// newUniverseLeafCache creates a new universe leaf cache.
+func newUniverseLeafCache() *universeLeafCache {
+
+	return &universeLeafCache{
+		leafCache: lru.NewCache[treeID, *cachedLeafKeys](
+			numCachedProofs,
+		),
+	}
+}
+
+// fetchLeafKeys reads the cached leaf keys for the given ID.
+func (u *universeLeafCache) fetchLeafKeys(id universe.Identifier,
+) []universe.LeafKey {
+
+	leafKeys, err := u.leafCache.Get(treeID(id.String()))
+	if err == nil {
+		log.Debugf("read leaf keys for %v from cache",
```

Comment:
nit: demote to trace to avoid spamming the log? As we'd expect to hit this quite often I think.